tensorflow_addons/metrics/multilabel_confusion_matrix.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements Multilabel confusion matrix scores."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.keras.metrics import Metric
+import numpy as np
+
+
+class MultiLabelConfusionMatrix(Metric):
+    def __init__(self,
+                 num_classes,
+                 name='Multilabel_confusion_matrix',
+                 dtype=tf.float32):
+        super(MultiLabelConfusionMatrix, self).__init__(name=name)
+        self.num_classes = num_classes
+        self.true_positives = self.add_weight(
+            'true_positives',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+        self.false_positives = self.add_weight(
+            'false_positives',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+        self.false_negatives = self.add_weight(
+            'false_negatives',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+        self.true_negatives = self.add_weight(
+            'true_negatives',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+
+    def update_state(self, y_true, y_pred):
+        y_true = tf.cast(y_true, tf.int32)
+        y_pred = tf.cast(y_pred, tf.int32)
+
+        # intermediate true positive calculation
+        true_positive_inter = y_true * y_pred
+        # true positives
+        tp = tf.math.count_nonzero(true_positive_inter, 0)
+        # predictions sum
+        pred_sum = tf.math.count_nonzero(y_pred, 0)
+        # true labels sum
+        true_sum = tf.math.count_nonzero(y_true, 0)
+        # false positives
+        print(tf.size(tp))
+        print(tp)
+        fp = pred_sum - tp
+        # false negatives
+        fn = true_sum - tp
+        # true negatives
+        print('in')
+        tn = y_true.get_shape()[0] - tp - fp - fn
```

Comment:
Correct me if I'm wrong, but won't `y_true.get_shape()[0]` give you the batch size? It should be `y_true.get_shape()[1]` right?

Reply:
It calculates true positives, false positives and false negatives first. To get the true negative value, it subtracts from the batch size. From there it extracts all the 4 values corresponding to each class.